android/src/main/java/software/solid/fluttervlcplayer/FlutterVideoView.java

```diff
@@ -100,46 +192,129 @@ public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
+                    case "stop":
+                        mediaPlayer.stop();
+                        break;
                 }
+
+                result.success(null);
+                break;
+
+            case "setPlaybackSpeed":
+
+                float playbackSpeed = Float.parseFloat((String) methodCall.argument("speed"));
+                mediaPlayer.setRate(playbackSpeed);
+
+                result.success(null);
+                break;
+
+            case "setTime":
+
+                long time = Long.parseLong((String) methodCall.argument("time"));
+                mediaPlayer.setTime(time);
+
+                result.success(null);
                 break;
         }
     }
 
     @Override
     public void onEvent(MediaPlayer.Event event) {
-        Map<String, String> resultMap = new HashMap<>();
+        HashMap<String, Object> eventObject = new HashMap<>();
 
         switch (event.type) {
-            case MediaPlayer.Event.Vout:
-                String aspectRatio;
+            case MediaPlayer.Event.Playing:
+                // Insert buffering=false event first:
+                eventObject.put("name", "buffering");
+                eventObject.put("value", false);
+                eventSink.success(eventObject.clone());
+                eventObject.clear();
+
+                // Now send playing info:
                 int height = 0;
                 int width = 0;
-                Media.VideoTrack currentVideoTrack = mediaPlayer.getCurrentVideoTrack();
+
+                Media.VideoTrack currentVideoTrack = (Media.VideoTrack) mediaPlayer.getMedia().getTrack(
+                    mediaPlayer.getVideoTrack()
+                );
                 if (currentVideoTrack != null) {
                     height = currentVideoTrack.height;
                     width = currentVideoTrack.width;
                 }
 
-                if (height != 0) {
-                    aspectRatio = String.valueOf(width / height);
-                    resultMap.put("aspectRatio", aspectRatio);
-                }
+                eventObject.put("name", "playing");
+                eventObject.put("value", true);
+                eventObject.put("ratio", height > 0 ? (double) width / (double) height : 0D);
+                eventObject.put("height", height);
+                eventObject.put("width", width);
+                eventObject.put("length", mediaPlayer.getLength());
+                eventSink.success(eventObject.clone());
+                break;
+
+            case MediaPlayer.Event.EndReached:
+                mediaPlayer.stop();
+                eventObject.put("name", "ended");
+                eventSink.success(eventObject);
+
+                eventObject.clear();
+                eventObject.put("name", "playing");
+                eventObject.put("value", false);
+                eventObject.put("reason", "EndReached");
+                eventSink.success(eventObject);
+
+            /*case MediaPlayer.Event.Opening:
```

Comment:
Do we need this commented code? Can you give a bit more context about it?

Reply:
@solid-software  Oh it was existing in the original fork. I forked the fork. Fixed the fork. Fixed a crash in the fork and submitted a PR. So some of this is left over.... Happy to fix, just not my code.

Reply:
Yeah this can be removed. It was based upon an incorrect assumption about how the library worked and I forgot to remove it.

Reply:
@solid-software resolved